service/src/main/kotlin/fi/espoo/evaka/specialdiet/SpecialDietQueries.kt

```diff
@@ -5,35 +5,50 @@
 package fi.espoo.evaka.specialdiet
 
 import fi.espoo.evaka.shared.ChildId
+import fi.espoo.evaka.shared.DaycareId
 import fi.espoo.evaka.shared.db.Database
+import java.time.LocalDate
+import org.jdbi.v3.core.mapper.Nested
 import org.jdbi.v3.core.mapper.PropagateNull
 
 data class MealTexture(@PropagateNull val id: Int?, val name: String)
 
 data class SpecialDiet(@PropagateNull val id: Int?, val abbreviation: String)
 
+data class NulledSpecialDiet(
+    val unitId: DaycareId,
+    val childId: ChildId,
+    @Nested("special_diet") val specialDiet: SpecialDiet,
+)
+
 /**
  * Sets child's special diets to null if there are some children whose special diet is not contained
  * in the new list. Returns a map from ChildId to SpecialDiet that was previously set for that
  * child.
  */
 fun Database.Transaction.resetSpecialDietsNotContainedWithin(
-    specialDietList: List<SpecialDiet>
-): Map<ChildId, SpecialDiet> {
+    today: LocalDate,
+    specialDietList: List<SpecialDiet>,
+): List<NulledSpecialDiet> {
     val newSpecialDietIds = specialDietList.map { it.id }
 
     val previousDiets =
         createQuery {
                 sql(
                     """
-SELECT child.id as child_id, special_diet.id, special_diet.abbreviation 
+SELECT
+    pl.unit_id,
+    child.id AS child_id,
+    special_diet.id AS special_diet_id,
+    special_diet.abbreviation AS special_diet_abbreviation
 FROM child 
-INNER JOIN special_diet ON child.diet_id = special_diet.id
+JOIN placement pl ON pl.child_id = child.id AND daterange(pl.start_date, pl.end_date, '[]') @> ${bind(today)}
```

Comment:
what if placement starts tomorrow and critical allergies have been already added?

Reply:
maybe lateral join with the earliest current/future placement could work 🤔

Reply:
Fixed